spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+
+### Ensuring hardware compatibility
+Immersive sessions will always be able to provide a `XRStationaryFrameOfReference`, but may not support other `XRFrameOfReference` types due to hardware limitations.  Developers are strongly encouraged to follow the spirit of [progressive enhanncement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement) and provide a reasonable fallback behavior if their desired `XRBoundedFrameOfReference` or `XRUnboundedFrameOfReference` is unavailable.  In many cases it will be adequate for this fallback to behave similarly to an inline preview experience.
+
+```js
+let xrSession = null;
+let xrFrameOfReference = null;
+
+function onSessionStarted(session) {
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'unbounded' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .catch(() => {
+    // Build an appropriate fallback experience if possible; perhaps similar to the inline version
+    xrSession.requestFrameOfReference({ type:'stationary', subtype:'eye-level' })
+    .then((frameOfReference) => {
+      xrFrameOfReference = frameOfReference;
+    });
+  })
+  .then(setupWebGLLayer)
+  .then(() => {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
+
+While many sites will be able to provide this fallback, for some sites this will not be possible.  Under these circumstances, it is instead preferable for session creation to reject rather than spin up immersive display/tracking systems only to immediately exit the session.
+
+```js
+function beginImmersiveSession() {
+  xrDevice.requestSession({ immersive: true,  requiredFrameOfReferenceType:'unbounded' })
+      .then(onSessionStarted)
+      .catch(err => {
+        // Error will indicate required frame of reference type unavailable
+      });
+}
+```
+
+### Floor Alignment Compatibility
+Some XR hardware with inside-out tracking has users establish "known spaces" that can be used to easily provide `XRBoundedFrameOfReference` and the `floor-level` subtype of `XRStationaryFrameOfReference`.  On inside-out XR hardware which does not intrinsically provide these known spaces, the User Agent may respond to the promise request by presenting UI for selecting a floor-aligned origin.  If the User Agent chooses not to do this, it will reject the request for floor aligned `XRFrameOfReference` objects with an error indicating the reason.  In response, a polyfill can be used to provide the same behavior by using another `XRFrameOfReference` type to draw UI for selecting a floor origin.  If anchors are available to the polyfill, it could then place an anchor at the user's selected floor point.
```

Comment:
What is the rationale for not requiring a US to provide a floor-aligned origin?  Seems like this greatly complicates user code?  Perhaps there's a way for the UA to say "it's just a guess" but at least be required to give a guess, so simple apps can just "get on with it"?

Reply:
We could do this, but I'd really like to get positive confirmation from all the hardware/UA folks that this is something they'd be willing to commit to.  Let's track this in a separate issue once the main PR is merged?

Reply:
Agreed - I can't see a UA doing any _worse_ than if an app ran with the default emulated floor height.  A UA should have access to either its own user-provided, platform-provided (or worst case, UA default) value for user height, which should be better than the app-provided emulated height, except in special cases where the app knows the user is sitting/kneeling.

It also benefits the WebXR ecosystem to get all standing apps that will guess at the floor anyway to just use a real `"floor-level"` frame of reference, since that enables best results on devices with real floor data or a convention to shift the XZ origin for standing experiences to an alternate location.

Reply:
@lincolnfrog Is this something Chrome would be comfortable doing on top of ARCore?
@rcabanier How about Helio?
@Artyom17 Oculus Browser?

(Also, general feedback from @rcarbanier and @artyom17 would be very appreciated!~)

Reply:
Helio would have no problem providing a "floor-level" frame of reference

Reply:
Ok, i'm going to make the change.  @artyom17, yell if that's a problem!

Reply:
@NellWaliczek I've talked to some people internally on the unbounded proposal and they believe this is a good approach and something we can provide. 
They did express some concern that this might be difficult to use for authors

Reply:
Hey Rik,
Good to hear that the design is implementable on Magic Leap!  I'm curious if you can share more details about the usability concern for authors?  I know that the lack of anchors in the current document iteration makes it a bit challenging to intuit the recommended unbounded usage pattern. Or is it something else?

Reply:
I suspect it is because we are used to hiding this complexity from the user.